core-primitives/attestation-handler/src/attestation_handler.rs

```diff
@@ -0,0 +1,638 @@
+		let ias_socket = self.ocall_api.get_ias_socket()?;
+
+		let mut rhs_vec: Vec<u8> = quote_nonce.rand.to_vec();
+		rhs_vec.extend(&quote_content);
+		let rhs_hash = rsgx_sha256_slice(&rhs_vec[..])?;
+		let lhs_hash = &qe_report.body.report_data.d[..32];
+
+		debug!("    [Enclave] rhs hash = {:02X}", rhs_hash.iter().format(""));
+		debug!("    [Enclave] lhs hash = {:02X}", lhs_hash.iter().format(""));
+
+		if rhs_hash != lhs_hash {
+			error!("    [Enclave] Quote is tampered!");
+			return Err(sgx_status_t::SGX_ERROR_UNEXPECTED)
+		}
+
+		let (attn_report, sig, cert) = self.get_report_from_intel(ias_socket, quote_content)?;
+		Ok((attn_report, sig, cert))
+	}
+
+	fn load_spid(filename: &str) -> SgxResult<sgx_spid_t> {
+		match io::read_to_string(filename).map(|contents| decode_spid(&contents)) {
+			Ok(r) => r,
+			Err(e) => {
+				error!("Failed to load SPID: {:?}", e);
+				Err(sgx_status_t::SGX_ERROR_UNEXPECTED)
+			},
+		}
+	}
+
+	fn get_ias_api_key() -> EnclaveResult<String> {
+		io::read_to_string(RA_API_KEY_FILE)
+			.map(|key| key.trim_end().to_owned())
+			.map_err(|e| EnclaveError::Other(e.into()))
+	}
+
+	pub fn create_ra_report_and_signature(
+		&self,
+		sign_type: sgx_quote_sign_type_t,
+		skip_ra: bool,
+	) -> EnclaveResult<(Vec<u8>, Vec<u8>)> {
+		let chain_signer = Ed25519Seal::unseal_from_static_file()?;
+		info!("[Enclave Attestation] Ed25519 pub raw : {:?}", chain_signer.public().0);
+
+		info!("    [Enclave] Generate keypair");
+		let ecc_handle = SgxEccHandle::new();
+		let _result = ecc_handle.open();
+		let (prv_k, pub_k) = ecc_handle.create_key_pair()?;
+		info!("    [Enclave] Generate ephemeral ECDSA keypair successful");
+		debug!("     pubkey X is {:02x}", pub_k.gx.iter().format(""));
+		debug!("     pubkey Y is {:02x}", pub_k.gy.iter().format(""));
+
+		let payload = if !skip_ra {
+			info!("    [Enclave] Create attestation report");
+			let (attn_report, sig, cert) =
+				match self.create_attestation_report(&chain_signer.public().0, sign_type) {
+					Ok(r) => r,
+					Err(e) => {
+						error!("    [Enclave] Error in create_attestation_report: {:?}", e);
+						return Err(e.into())
+					},
+				};
+			println!("    [Enclave] Create attestation report successful");
+			debug!("              attn_report = {:?}", attn_report);
+			debug!("              sig         = {:?}", sig);
+			debug!("              cert        = {:?}", cert);
+
+			// concat the information
+			attn_report + "|" + &sig + "|" + &cert
+		} else {
+			Default::default()
+		};
+
+		// generate an ECC certificate
+		info!("    [Enclave] Generate ECC Certificate");
+		let (key_der, cert_der) = match cert::gen_ecc_cert(payload, &prv_k, &pub_k, &ecc_handle) {
+			Ok(r) => r,
+			Err(e) => {
+				error!("    [Enclave] gen_ecc_cert failed: {:?}", e);
+				return Err(e.into())
+			},
+		};
+
+		let _ = ecc_handle.close();
+		info!("    [Enclave] Generate ECC Certificate successful");
+		Ok((key_der, cert_der))
+	}
+
+	pub unsafe fn perform_ra(
```

Comment:
This function interface should also be of Rust type, not unsafe C. The E-call implementation should be responsible for converting the parameters between the unsafe C interface and Rust. Our Rust components should only have Rust function interfaces.